### PR TITLE
exposing layout op at direct python binding

### PR DESCRIPTION
**Pull request by jjsjann123**
## Stacked PRs

#5230 moe layer with nvfp4 grouped_mm
#5345 exposing layout op at direct python binding  <-- this PR
#5198 refactor number of groups in layout op
#5174 allow layout op in automatic scheduler

## This PR

Expose layout op at python direct binding.
Added nvfp4 grouped gemm in python test.

Minor fixes:

1. ~Added support of allocation domain for output of layout op in concretization pass to maintain the dependency on padded allocation domain to its logical domain.~ No longer needed, handled in https://github.com/NVIDIA/Fuser/pull/5384
2. Skipped validation for `setAllocationDomain`
3. updated reference implementation to match the math order in nvfuser decomposed nvfp4 quantization.

TODO:

python tests requires IdModel Indexer in order to work. See issue #5200, as well as suggested WAR in https://github.com/NVIDIA/Fuser/issues/5200#issuecomment-3319839361

### CODE REVIEW

**Comment by jjsjann123** [comment]
!test

File: csrc/dynamic_transform.cpp [hunk 1]
```diff
@@ -848,6 +849,22 @@ void DynamicTransformConcretizer::concretize() {
...
+  for (auto* layout_op :
+       ir_utils::filterByType<PreprocessGroupedMatmulInputSf>(exprs)) {
+    auto* out_tv = layout_op->out()->as<TensorView>();
+    std::vector<IterDomain*> logical_dom =
+        TensorDomain::noReductions(out_tv->getLogicalDomain());
+    std::vector<IterDomain*> alloc_dom = layoutAllocationDomain(
+        logical_dom, layout_op->g(), layout_op->layout());
+    // skip validation because allocation domain doesn't converge to logical
+    // domain.
+    out_tv->domain()->setAllocationDomain(alloc_dom, true, true);
```
Comment (jjsjann123) [inline csrc/dynamic_transform.cpp]: tagging @jacobhinkle since I'm touching concretization.

File: python/python_direct/ops.cpp [hunk 2]
```diff
@@ -3188,6 +3188,40 @@ Notes
...
+    A 1D tensor with length as `number of groups`.
+    Its value notes the offsets of the starting token in each group for the input tensor view
+output_offsets: TensorView
+    A 1D tensor with length as `number of groups`.
+    Its value notes the offsets of the starting token in each group for the output tensor view.
+
+Returns
+-------
+TensorView
+    A tensor with proper swizzle & padding  in memory. Note that the actual padding in buffer is not represented by the size/stride of the output tensor.
```
Comment (rdspring1) [inline python/python_direct/ops.cpp]:
```suggestion
    A tensor with proper swizzle & padding in memory. Note that the actual padding in buffer is not represented by the size/stride of the output tensor.
```

File: python/python_direct/ops.cpp [hunk 3]
```diff
@@ -3188,6 +3188,40 @@ Notes
...
+            input,
+            input_offsets,
+            output_offsets,
+            BlockScalingFactorLayout::Block128x4);
+      },
+      py::arg("input"),
+      py::arg("input_offsets"),
+      py::arg("output_offsets"),
+      R"(
+Layout operation to apply per group swizzle & padding for grouped matmul block scaling factor for activation.
```
Comment (rdspring1) [inline python/python_direct/ops.cpp]:
```suggestion
Layout operation to apply per group swizzle & padding to the block scaling factor of the input activations to grouped matmul.
```

File: tests/python/direct_utils/narrow_precision.py [hunk 4]
```diff
@@ -137,11 +137,11 @@ def pytorch_nvfp4_quantize(a, a_global_scale):
...
     scaled_block_scale_fp32 = block_scale_fp32 * a_global_scale
-    scaled_block_scale_fp8 = torch.clamp(
+    scaled_block_scale_fp32 = torch.clamp(
         scaled_block_scale_fp32, min=FLOAT8_E4M3_EPS, max=FLOAT8_E4M3_MAX
-    ).to(torch.float8_e4m3fn)
-    scaled_block_scale_fp8_fp32 = scaled_block_scale_fp8.to(torch.float)
-    total_scale = scaled_block_scale_fp8_fp32 / a_global_scale
+    )
+    scaled_block_scale_fp8 = scaled_block_scale_fp32.to(torch.float8_e4m3fn)
+    total_scale = scaled_block_scale_fp32 / a_global_scale
```
Comment (rdspring1) [inline tests/python/direct_utils/narrow_precision.py:144]: 👍🏼

File: csrc/dynamic_transform.cpp (hunk 1 above)
Comment (rdspring1) [inline csrc/dynamic_transform.cpp]:
```suggestion
    out_tv->domain()->setAllocationDomain(alloc_dom, /*new_contiguity=*/true, /*skip_validation=*/true);
```

File: tests/python/direct/test_narrow_precision.py [hunk 5]
```diff
@@ -271,3 +269,188 @@ def nvfuser_fusion_id0(fd: FusionDefinition) -> None:
         )
 
     assert torch.allclose(o_decomposed_ref, o[0], atol=1e-2, rtol=1e-2)
+
+
+# TODO: update reference implementation to support padding on k
+@pytest.mark.skipif(
+    is_pre_blackwell(), reason="Only supported on blackwell and newer devices."
+)
```
Comment (rdspring1) [inline tests/python/direct/test_narrow_precision.py]: @pytest.mark.skipif(
    not microarchitecture_is_pre(12), reason="Does not support blackwell compute 12.0."
)

File: csrc/dynamic_transform.cpp [hunk 6]
```diff
@@ -848,6 +849,22 @@ void DynamicTransformConcretizer::concretize() {
     OptOutMutator::dispatchMutate(stmt);
   }
 
+  // OptOutMutator only updates ID, but not exprs on its extent. Concretization
+  // on the allocation domain of layout op output needs to be manually replaced,
+  // because they are not connect by ID ops.
```
Comment (jacobhinkle) [inline csrc/dynamic_transform.cpp]: We do traverse all attributes in the `stmt` loop above, so it should hit all IDs in the `TensorDomain` in topo order. Did you try relaxing the condition at line 845 and then doing this in `handle(PreprocessGroupedMatmulInputSf)`?

File: csrc/ir/nodes.cpp [hunk 7]
```diff
@@ -4082,11 +4082,14 @@ void TensorDomain::setAlternateLoopDomain(
 
 void TensorDomain::setAllocationDomain(
     std::vector<IterDomain*> new_allocation_domain,
-    std::vector<std::optional<bool>> new_contiguity) {
+    std::vector<std::optional<bool>> new_contiguity,
+    bool skip_validation) {
```
Comment (jjsjann123) [inline csrc/ir/nodes.cpp:4087]: I wonder if we still need this, since we already marked allocation domain as symbolic now.

File: csrc/dynamic_transform.cpp (hunk 6 above)
Comment (jjsjann123) [reply csrc/dynamic_transform.cpp]: > so it should hit all IDs in the TensorDomain in topo order.

I thought we only traverse to loop, since https://github.com/NVIDIA/Fuser/blob/de077d73d507558a742e813b01d7d6f59a9f0a15/csrc/iter_visitor.cpp#L72-L75

With the new layout op, its allocation domain is not connected to logical, nor is it on the path to loop. It could look like this vvv, where it's extent is directly computed (symbolically).

```
(gdb) p consumer->printTransforms()
 logical domain : (iS54{i0}, iS55{9})
 allocation domain : (iS56{( i0 + ( i3 * 127 ) )}, iS57{( ( 12 / 4 ) * 4 )})
 contiguity: t t
 loop domain : (iS54{i0}, iS55{9})
```

**Comment by jjsjann123** [comment]
!test

**Comment by jjsjann123** [comment]
!test

File: tests/python/direct/test_narrow_precision.py [hunk 8]
```diff
@@ -278,3 +276,191 @@ def nvfuser_fusion_id0(fd: FusionDefinition) -> None:
...
+        scale2,
+        mat2_gs,
+        problem_sizes,
+        offsets,
+        blockscale_offsets,
+    ]
+
+    # FIXME: force indexing to use IdModel indexer to avoid indexing error.
+    # see issue: https://github.com/NVIDIA/Fuser/issues/5200
+    with set_env(NVFUSER_ENABLE="id_model(all)"):
```
Comment (jjsjann123) [inline tests/python/direct/test_narrow_precision.py]: this doesn't seem to work when we run multiple tests together, since we cache the env_variable. I need to change this to the scope of this file maybe... or just skip this test.

File: csrc/dynamic_transform.cpp (hunk 6 above)
Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp]: That's right. Instead I'm going to try updating IterVisitor to traverse allIDs in the TensorDomain instead.

Comment (jacobhinkle) [reply csrc/dynamic_transform.cpp]: Confirmed that if we merge #5384, then we can revert the changes in this file and the included test passes.

Comment (jjsjann123) [reply csrc/dynamic_transform.cpp]: Awesome. Thanks for patching that.

**Comment by jjsjann123** [comment]
Had to rebase & force push to avoid resolving all conflicts from the base change.

File: csrc/dynamic_transform.cpp (hunk 6 above)
Comment (jjsjann123) [reply csrc/dynamic_transform.cpp]: reverted. Thanks a lot for helping out.

**Comment by jjsjann123** [comment]
!test

**Comment by jjsjann123** [comment]
everything looks good, except the env var thing that's not working properly. I might change that to a standalone python test, just so it's not going to mess with the other tests.

**Comment by jjsjann123** [comment]
!test

File: tests/python/direct_utils/narrow_precision.py (hunk 4 above)
Comment (jjsjann123) [reply tests/python/direct_utils/narrow_precision.py:144]: @rdspring1 This apparently is triggering numerical issues for tests.python.direct.test_cutlass_nvfp4_gemm.test_nvfp4_gemm_epilogue

https://gitlab-master.nvidia.com/dl/pytorch/fuser-gh-mirror/-/pipelines/36930137

I wonder why do we check the result and not using the assert close instead?

**Comment by jjsjann123** [comment]
!test
